runtime error: invalid memory address or nil pointer dereference

Not sure what I've done wrong but I've been met with this error when trying to do an initial setup of notifiarr on unraid.
```
[ERROR] 2022/08/20 13:17:20 Go Panic! This is a bug please report it on github: https://github.com/Notifiarr/notifiarr/issues/new
runtime error: invalid memory address or nil pointer dereference
goroutine 1 [running]:
runtime/debug.Stack()
        runtime/debug/stack.go:24 +0x65
main.main.func1()
        github.com/Notifiarr/notifiarr/main.go:21 +0x3d
panic({0xc64ba0, 0x1555340})
        runtime/panic.go:884 +0x212
github.com/Notifiarr/notifiarr/pkg/services.(*Config).collectDownloadApps(0xc0002c46e0, {0x159ce50?, 0xf?, 0xd1e7b0?})
        github.com/Notifiarr/notifiarr/pkg/services/apps.go:198 +0x36b
github.com/Notifiarr/notifiarr/pkg/services.(*Config).collectApps(0xc0002c46e0)
        github.com/Notifiarr/notifiarr/pkg/services/apps.go:18 +0x9f
github.com/Notifiarr/notifiarr/pkg/services.(*Config).Setup(0xc0002c46e0, {0x0?, 0x0, 0x0})
        github.com/Notifiarr/notifiarr/pkg/services/services.go:29 +0x85
github.com/Notifiarr/notifiarr/pkg/configfile.(*Config).Get(0xc0002e6140, 0xc0002ec090)
        github.com/Notifiarr/notifiarr/pkg/configfile/config.go:150 +0x39e
github.com/Notifiarr/notifiarr/pkg/client.(*Client).loadConfiguration(0xc0002c4780)
        github.com/Notifiarr/notifiarr/pkg/client/start.go:173 +0x1a5
github.com/Notifiarr/notifiarr/pkg/client.(*Client).start(0xc0002c4780)
        github.com/Notifiarr/notifiarr/pkg/client/start.go:106 +0x36
github.com/Notifiarr/notifiarr/pkg/client.Start()
        github.com/Notifiarr/notifiarr/pkg/client/start.go:101 +0xda
main.main()
        github.com/Notifiarr/notifiarr/main.go:25 +0x6a
[ERROR] 2022/08/20 13:20:43 Go Panic! This is a bug please report it on github: https://github.com/Notifiarr/notifiarr/issues/new
runtime error: invalid memory address or nil pointer dereference
goroutine 1 [running]:
runtime/debug.Stack()
        runtime/debug/stack.go:24 +0x65
main.main.func1()
        github.com/Notifiarr/notifiarr/main.go:21 +0x3d
panic({0xc64ba0, 0x1555340})
        runtime/panic.go:884 +0x212
github.com/Notifiarr/notifiarr/pkg/services.(*Config).collectDownloadApps(0xc0001e4000, {0x159ce50?, 0xc0000de5c0?, 0x3c?})
        github.com/Notifiarr/notifiarr/pkg/services/apps.go:198 +0x36b
github.com/Notifiarr/notifiarr/pkg/services.(*Config).collectApps(0xc0001e4000)
        github.com/Notifiarr/notifiarr/pkg/services/apps.go:18 +0x9f
github.com/Notifiarr/notifiarr/pkg/services.(*Config).Setup(0xc0001e4000, {0x0?, 0x0, 0x0})
        github.com/Notifiarr/notifiarr/pkg/services/services.go:29 +0x85
github.com/Notifiarr/notifiarr/pkg/configfile.(*Config).Get(0xc0000f5180, 0xc0001bc120)
        github.com/Notifiarr/notifiarr/pkg/configfile/config.go:150 +0x39e
github.com/Notifiarr/notifiarr/pkg/client.(*Client).loadConfiguration(0xc0001e40a0)
        github.com/Notifiarr/notifiarr/pkg/client/start.go:173 +0x1a5
github.com/Notifiarr/notifiarr/pkg/client.(*Client).start(0xc0001e40a0)
        github.com/Notifiarr/notifiarr/pkg/client/start.go:106 +0x36
github.com/Notifiarr/notifiarr/pkg/client.Start()
        github.com/Notifiarr/notifiarr/pkg/client/start.go:101 +0xda
main.main()
        github.com/Notifiarr/notifiarr/main.go:25 +0x6a
```

Sorry if it's something obvious that I'm overlooking.

--- a/pkg/client/start.go
+++ b/pkg/client/start.go
@@ -299,7 +299,7 @@ func (c *Client) reloadConfiguration(event website.EventType, source string) err
 	c.Logger.SetupLogging(c.Config.LogConfig)
 	clientInfo := c.configureServices()
 	c.setupMenus(clientInfo)
-	c.Print(" 🔄 Configuration Reloaded! Config File:", c.Flags.ConfigFile)
+	c.Print(" 🌀 Configuration Reloaded! Config File:", c.Flags.ConfigFile)
 
 	if err = ui.Notify("Configuration Reloaded! Config File: %s", c.Flags.ConfigFile); err != nil {
 		c.Errorf("Creating Toast Notification: %v", err)

--- a/pkg/logs/logs.go
+++ b/pkg/logs/logs.go
@@ -20,6 +20,7 @@ import (
 	homedir "github.com/mitchellh/go-homedir"
 	"golift.io/rotatorr"
 	"golift.io/rotatorr/timerotator"
+	"golift.io/version"
 )
 
 // Logger provides some methods with baked in assumptions.
@@ -165,7 +166,8 @@ func (l *Logger) CapturePanic() {
 	if r := recover(); r != nil {
 		ui.ShowConsoleWindow()
 		l.ErrorLog.Output(callDepth, //nolint:errcheck
-			fmt.Sprintf("Go Panic! %s\n%v\n%s", mnd.BugIssue, r, string(debug.Stack())))
+			fmt.Sprintf("Go Panic! %s\n%s-%s %s %v\n%s", mnd.BugIssue,
+				version.Version, version.Revision, version.Branch, r, string(debug.Stack())))
 		panic(r)
 	}
 }

--- a/pkg/services/apps.go
+++ b/pkg/services/apps.go
@@ -24,7 +24,7 @@ func (c *Config) collectApps() []*Service {
 
 func (c *Config) collectLidarrApps(svcs []*Service) []*Service {
 	for _, app := range c.Apps.Lidarr {
-		if !app.Enabled() {
+		if !app.Enabled() || app.Name == "" {
 			continue
 		}
 
@@ -51,7 +51,7 @@ func (c *Config) collectLidarrApps(svcs []*Service) []*Service {
 
 func (c *Config) collectProwlarrApps(svcs []*Service) []*Service {
 	for _, app := range c.Apps.Prowlarr {
-		if !app.Enabled() {
+		if !app.Enabled() || app.Name == "" {
 			continue
 		}
 
@@ -78,7 +78,7 @@ func (c *Config) collectProwlarrApps(svcs []*Service) []*Service {
 
 func (c *Config) collectRadarrApps(svcs []*Service) []*Service {
 	for _, app := range c.Apps.Radarr {
-		if !app.Enabled() {
+		if !app.Enabled() || app.Name == "" {
 			continue
 		}
 
@@ -105,7 +105,7 @@ func (c *Config) collectRadarrApps(svcs []*Service) []*Service {
 
 func (c *Config) collectReadarrApps(svcs []*Service) []*Service {
 	for _, app := range c.Apps.Readarr {
-		if !app.Enabled() {
+		if !app.Enabled() || app.Name == "" {
 			continue
 		}
 
@@ -132,7 +132,7 @@ func (c *Config) collectReadarrApps(svcs []*Service) []*Service {
 
 func (c *Config) collectSonarrApps(svcs []*Service) []*Service {
 	for _, app := range c.Apps.Sonarr {
-		if !app.Enabled() {
+		if !app.Enabled() || app.Name == "" {
 			continue
 		}
 
@@ -157,10 +157,11 @@ func (c *Config) collectSonarrApps(svcs []*Service) []*Service {
 	return svcs
 }
 
-func (c *Config) collectDownloadApps(svcs []*Service) []*Service { //nolint:funlen,cyclop,gocognit
+//nolint:funlen,cyclop,gocognit,gocyclo // split this one up.
+func (c *Config) collectDownloadApps(svcs []*Service) []*Service {
 	// Deluge instanceapp.
 	for _, app := range c.Apps.Deluge {
-		if app.Timeout.Duration < 0 {
+		if app == nil || app.Config == nil || app.Name == "" || app.Timeout.Duration < 0 {
 			continue
 		}
 
@@ -184,7 +185,7 @@ func (c *Config) collectDownloadApps(svcs []*Service) []*Service { //nolint:funl
 
 	// NZBGet instances.
 	for _, app := range c.Apps.NZBGet {
-		if app.Timeout.Duration < 0 {
+		if app == nil || app.Config == nil || app.Name == "" || app.Timeout.Duration < 0 {
 			continue
 		}
 
@@ -217,7 +218,7 @@ func (c *Config) collectDownloadApps(svcs []*Service) []*Service { //nolint:funl
 
 	// Qbittorrent instanceapp.
 	for _, app := range c.Apps.Qbit {
-		if app.Timeout.Duration < 0 {
+		if app == nil || app.Config == nil || app.Name == "" || app.Timeout.Duration < 0 {
 			continue
 		}
 
@@ -241,7 +242,7 @@ func (c *Config) collectDownloadApps(svcs []*Service) []*Service { //nolint:funl
 
 	// rTorrent instanceapp.
 	for _, app := range c.Apps.Rtorrent {
-		if app.Timeout.Duration < 0 {
+		if app == nil || app.Name == "" || app.Timeout.Duration < 0 {
 			continue
 		}
 
@@ -265,7 +266,7 @@ func (c *Config) collectDownloadApps(svcs []*Service) []*Service { //nolint:funl
 
 	// SabNBZd instanceapp.
 	for _, app := range c.Apps.SabNZB {
-		if app.Timeout.Duration < 0 {
+		if app == nil || app.Name == "" || app.Timeout.Duration < 0 {
 			continue
 		}
 
@@ -292,37 +293,36 @@ func (c *Config) collectDownloadApps(svcs []*Service) []*Service { //nolint:funl
 
 func (c *Config) collectTautulliApp(svcs []*Service) []*Service {
 	// Tautulli instance (1).
-	if app := c.Apps.Tautulli; app != nil && app.URL != "" && app.Name != "" {
-		if app.Timeout.Duration < 0 {
-			return svcs
-		}
-
-		interval := app.Interval
-		if interval.Duration == 0 {
-			interval.Duration = DefaultCheckInterval
-		}
-
-		svcs = append(svcs, &Service{
-			Name:     app.Name,
-			Type:     CheckHTTP,
-			Value:    app.URL + "/api/v2?cmd=status&apikey=" + app.APIKey,
-			Expect:   "200",
-			Timeout:  app.Timeout,
-			Interval: interval,
-			validSSL: app.VerifySSL,
-		})
+	app := c.Apps.Tautulli
+	if app == nil || app.URL == "" || app.Name == "" || app.Timeout.Duration < 0 {
+		return svcs
 	}
+
+	interval := app.Interval
+	if interval.Duration == 0 {
+		interval.Duration = DefaultCheckInterval
+	}
+
+	svcs = append(svcs, &Service{
+		Name:     app.Name,
+		Type:     CheckHTTP,
+		Value:    app.URL + "/api/v2?cmd=status&apikey=" + app.APIKey,
+		Expect:   "200",
+		Timeout:  app.Timeout,
+		Interval: interval,
+		validSSL: app.VerifySSL,
+	})
 
 	return svcs
 }
 
-func (c *Config) collectMySQLApps(svcs []*Service) []*Service {
+func (c *Config) collectMySQLApps(svcs []*Service) []*Service { //nolint:cyclop
 	if c.Plugins == nil {
 		return svcs
 	}
 
 	for _, plugin := range c.Plugins.MySQL {
-		if plugin.Timeout.Duration < 0 {
+		if plugin.Host == "" || plugin.Timeout.Duration < 0 {
 			continue
 		} else if plugin.Timeout.Duration == 0 {
 			plugin.Timeout.Duration = DefaultTimeout

--- a/pkg/services/interface.go
+++ b/pkg/services/interface.go
@@ -83,7 +83,7 @@ func (c *Config) runChecks(forceAll bool) bool {
 }
 
 func (c *Config) updateStatesOnSite(force bool) {
-	if !force && time.Since(c.lastUpdate) < time.Hour {
+	if !force && time.Since(c.lastUpdate) < time.Hour || len(c.services) == 0 {
 		return
 	}
 
@@ -104,7 +104,6 @@ func (c *Config) updateStatesOnSite(force bool) {
 	}
 
 	if len(values) == 0 {
-		c.Printf("No service states to send updates to website? Weird.")
 		return
 	}
 

--- a/pkg/triggers/filewatch/filewatch.go
+++ b/pkg/triggers/filewatch/filewatch.go
@@ -168,16 +168,14 @@ func (c *cmd) collectFileTails(tails []*WatchFile) ([]reflect.SelectCase, *time.
 
 //nolint:cyclop
 func (c *cmd) tailFiles(cases []reflect.SelectCase, tails []*WatchFile, ticker *time.Ticker) {
-	defer func() {
-		defer c.CapturePanic()
-		ticker.Stop()
-		c.Printf("==> All file watchers stopped.")
-		close(c.stopWatcher) // signal we're done.
-	}()
+	limitChan, limitTicker := getLimiter()
+	defer c.closeWatcher(limitChan, limitTicker, ticker)
 
 	var died bool
 
 	for {
+		<-limitChan
+
 		idx, data, running := reflect.Select(cases)
 		item := tails[idx]
 
@@ -219,6 +217,42 @@ func (c *cmd) tailFiles(cases []reflect.SelectCase, tails []*WatchFile, ticker *
 			exp.FileWatcher.Add(item.Path+" Bytes", int64(len(line.Text)))
 		}
 	}
+}
+
+// closeWatcher is defered when the watcher dies.
+func (c *cmd) closeWatcher(limiter chan time.Time, tickers ...*time.Ticker) {
+	defer c.CapturePanic()
+
+	for _, ticker := range tickers {
+		ticker.Stop()
+	}
+
+	c.Printf("==> All file watchers stopped.")
+	close(limiter)
+	close(c.stopWatcher) // signal we're done.
+}
+
+// getLimiter returns a channel that limits requests to 1 per second, with a burst rate of 10.
+func getLimiter() (chan time.Time, *time.Ticker) {
+	const (
+		burstRate  = 10
+		requestPer = time.Second
+	)
+
+	limiter := time.NewTicker(requestPer)
+
+	burstyLimiter := make(chan time.Time, burstRate)
+	for i := 0; i < 3; i++ {
+		burstyLimiter <- time.Now()
+	}
+
+	go func() {
+		for t := range limiter.C {
+			burstyLimiter <- t
+		}
+	}()
+
+	return burstyLimiter, limiter
 }
 
 func (c *cmd) fileWatcherTicker(died bool) bool {

--- a/pkg/triggers/filewatch/filewatch.go
+++ b/pkg/triggers/filewatch/filewatch.go
@@ -21,6 +21,9 @@ var ErrInvalidRegexp = fmt.Errorf("invalid regexp")
 const (
 	maxRetries    = 6
 	retryInterval = 10 * time.Second
+	specialCase   = 2           // We have two special channels in our select cases.
+	burstRate     = 10          // burst to this many 'matches' before throttling.
+	requestPer    = time.Second // 1 request per this time period allowed + burst rate.
 )
 
 type cmd struct {
@@ -144,6 +147,7 @@ func (c *cmd) collectFileTails(tails []*WatchFile) ([]reflect.SelectCase, *time.
 	cases := make([]reflect.SelectCase, len(tails))
 
 	for idx, item := range tails {
+		// If you add more special cases here, increment specialCase.
 		if idx == 0 { // 0 is skipped (see above), and used as an internal I/O channel
 			cases[idx] = reflect.SelectCase{Dir: reflect.SelectRecv, Chan: reflect.ValueOf(c.addWatcher)}
 			continue
@@ -166,40 +170,28 @@ func (c *cmd) collectFileTails(tails []*WatchFile) ([]reflect.SelectCase, *time.
 	return cases, ticker
 }
 
-//nolint:cyclop
 func (c *cmd) tailFiles(cases []reflect.SelectCase, tails []*WatchFile, ticker *time.Ticker) {
-	limitChan, limitTicker := getLimiter()
-	defer c.closeWatcher(limitChan, limitTicker, ticker)
+	closeLimiter := make(chan struct{})
+	defer c.closeWatcher(closeLimiter, ticker)
+	limitCh := getLimiter(closeLimiter)
 
 	var died bool
 
 	for {
-		<-limitChan
+		<-limitCh
 
 		idx, data, running := reflect.Select(cases)
 		item := tails[idx]
 
 		switch {
 		case !running && idx == 0:
-			return // main channel closed, bail out.
+			if len(cases) <= specialCase {
+				return // all channels are now closed, bail out.
+			}
 		case !running:
 			tails = append(tails[:idx], tails[idx+1:]...) // The channel was closed? okay, remove it.
 			cases = append(cases[:idx], cases[idx+1:]...)
-
-			// This routines runs the Stop method on the tail.
-			// If that returns an error, it means it died.
-			// If that does not return an error, it means Stop was already called.
-			if err := item.deactivate(); err != nil {
-				c.Errorf("No longer watching file (channel closed): %s: %v", item.Path, err)
-				exp.FileWatcher.Add(item.Path+" Errors", 1)
-				died = true //nolint:wsl
-			} else {
-				c.Printf("==> No longer watching file (channel closed): %s", item.Path)
-			}
-
-			if len(cases) < 1 {
-				return
-			}
+			died = c.killWatcher(item)
 		case idx == 1:
 			died = c.fileWatcherTicker(died)
 		case data.IsNil(), data.IsZero(), !data.Elem().CanInterface():
@@ -220,39 +212,58 @@ func (c *cmd) tailFiles(cases []reflect.SelectCase, tails []*WatchFile, ticker *
 }
 
 // closeWatcher is defered when the watcher dies.
-func (c *cmd) closeWatcher(limiter chan time.Time, tickers ...*time.Ticker) {
+func (c *cmd) closeWatcher(closeLimiter chan struct{}, ticker *time.Ticker) {
 	defer c.CapturePanic()
-
-	for _, ticker := range tickers {
-		ticker.Stop()
-	}
-
+	closeLimiter <- struct{}{}
+	<-closeLimiter // wait for it to close.
+	ticker.Stop()
 	c.Printf("==> All file watchers stopped.")
-	close(limiter)
 	close(c.stopWatcher) // signal we're done.
 }
 
+// killWatcher runs the Stop method on the tail.
+// If that returns an error, it means it died.
+// If that does not return an error, it means Stop was already called.
+func (c *cmd) killWatcher(item *WatchFile) bool {
+	if err := item.deactivate(); err != nil {
+		c.Errorf("No longer watching file (channel closed): %s: %v", item.Path, err)
+		exp.FileWatcher.Add(item.Path+" Errors", 1)
+
+		return true
+	}
+
+	c.Printf("==> No longer watching file (channel closed): %s", item.Path)
+
+	return false
+}
+
 // getLimiter returns a channel that limits requests to 1 per second, with a burst rate of 10.
-func getLimiter() (chan time.Time, *time.Ticker) {
-	const (
-		burstRate  = 10
-		requestPer = time.Second
-	)
+func getLimiter(done chan struct{}) chan time.Time {
+	limitTicker := time.NewTicker(requestPer)
+	limitCh := make(chan time.Time, burstRate)
 
-	limiter := time.NewTicker(requestPer)
-
-	burstyLimiter := make(chan time.Time, burstRate)
-	for i := 0; i < 3; i++ {
-		burstyLimiter <- time.Now()
+	for i := 0; i < burstRate; i++ {
+		limitCh <- time.Now()
 	}
 
 	go func() {
-		for t := range limiter.C {
-			burstyLimiter <- t
+		defer func() {
+			limitTicker.Stop()
+			close(limitCh)
+			close(done)
+		}()
+
+		for {
+			select {
+			case <-done:
+				return
+			case t := <-limitTicker.C:
+				limitCh <- t
+			}
 		}
 	}()
 
-	return burstyLimiter, limiter
+	return limitCh
 }
 
 func (c *cmd) fileWatcherTicker(died bool) bool {

--- a/pkg/triggers/filewatch/filewatch.go
+++ b/pkg/triggers/filewatch/filewatch.go
@@ -14,16 +14,17 @@ import (
 	"github.com/Notifiarr/notifiarr/pkg/triggers/common"
 	"github.com/Notifiarr/notifiarr/pkg/website"
 	"github.com/nxadm/tail"
+	"github.com/nxadm/tail/ratelimiter"
 )
 
 var ErrInvalidRegexp = fmt.Errorf("invalid regexp")
 
 const (
-	maxRetries    = 6
-	retryInterval = 10 * time.Second
-	specialCase   = 2           // We have two special channels in our select cases.
-	burstRate     = 10          // burst to this many 'matches' before throttling.
-	requestPer    = time.Second // 1 request per this time period allowed + burst rate.
+	maxRetries    = 6                                  // how many times to retry watching a file.
+	retryInterval = 10 * time.Second                   // how often channels are checked for being closed.
+	specialCase   = 2                                  // We have two special channels in our select cases.
+	burstRate     = 6                                  // burst to this many 'matches' before throttling.
+	requestPer    = time.Second + 500*time.Millisecond // 1 request per this time period allowed + burst rate.
 )
 
 type cmd struct {
@@ -128,6 +129,7 @@ func (w *WatchFile) setup(logger *log.Logger) error {
 		CompleteLines: true,
 		Location:      &tail.SeekInfo{Whence: io.SeekEnd},
 		Logger:        logger,
+		RateLimiter:   ratelimiter.NewLeakyBucket(burstRate, requestPer),
 	})
 	if err != nil {
 		exp.FileWatcher.Add(w.Path+" Errors", 1)
@@ -171,15 +173,16 @@ func (c *cmd) collectFileTails(tails []*WatchFile) ([]reflect.SelectCase, *time.
 }
 
 func (c *cmd) tailFiles(cases []reflect.SelectCase, tails []*WatchFile, ticker *time.Ticker) {
-	closeLimiter := make(chan struct{})
-	defer c.closeWatcher(closeLimiter, ticker)
-	limitCh := getLimiter(closeLimiter)
+	defer func() {
+		defer c.CapturePanic()
+		ticker.Stop()
+		c.Printf("==> All file watchers stopped.")
+		close(c.stopWatcher) // signal we're done.
+	}()
 
 	var died bool
 
 	for {
-		<-limitCh
-
 		idx, data, running := reflect.Select(cases)
 		item := tails[idx]
 
@@ -211,16 +214,6 @@ func (c *cmd) tailFiles(cases []reflect.SelectCase, tails []*WatchFile, ticker *
 	}
 }
 
-// closeWatcher is defered when the watcher dies.
-func (c *cmd) closeWatcher(closeLimiter chan struct{}, ticker *time.Ticker) {
-	defer c.CapturePanic()
-	closeLimiter <- struct{}{}
-	<-closeLimiter // wait for it to close.
-	ticker.Stop()
-	c.Printf("==> All file watchers stopped.")
-	close(c.stopWatcher) // signal we're done.
-}
-
 // killWatcher runs the Stop method on the tail.
 // If that returns an error, it means it died.
 // If that does not return an error, it means Stop was already called.
@@ -237,35 +230,7 @@ func (c *cmd) killWatcher(item *WatchFile) bool {
 	return false
 }
 
-// getLimiter returns a channel that limits requests to 1 per second, with a burst rate of 10.
-func getLimiter(done chan struct{}) chan time.Time {
-	limitTicker := time.NewTicker(requestPer)
-	limitCh := make(chan time.Time, burstRate)
-
-	for i := 0; i < burstRate; i++ {
-		limitCh <- time.Now()
-	}
-
-	go func() {
-		defer func() {
-			limitTicker.Stop()
-			close(limitCh)
-			close(done)
-		}()
-
-		for {
-			select {
-			case <-done:
-				return
-			case t := <-limitTicker.C:
-				limitCh <- t
-			}
-		}
-	}()
-
-	return limitCh
-}
-
+// fileWatcherTicker checks if a file watcher died and needs to be restarted.
 func (c *cmd) fileWatcherTicker(died bool) bool {
 	if !died {
 		return false

